src/vm/castcache.cpp

```diff
@@ -0,0 +1,275 @@
+
+#if !defined(DACCESS_COMPILE) && !defined(CROSSGEN_COMPILE)
+
+OBJECTHANDLE CastCache::s_cache = NULL;
+
+BASEARRAYREF CastCache::CreateCastCache(DWORD size)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    BASEARRAYREF table = NULL;
+
+    try
+    {
+        table = (BASEARRAYREF)AllocatePrimitiveArray(CorElementType::ELEMENT_TYPE_I8, (size + 1) * sizeof(CastCacheEntry) / sizeof(INT64));
+    }
+    catch (OutOfMemoryException)
+    {
+        // try a small cache
+        size = INITIAL_CACHE_SIZE;
+        try
+        {
+            table = (BASEARRAYREF)AllocatePrimitiveArray(CorElementType::ELEMENT_TYPE_I8, (size + 1) * sizeof(CastCacheEntry) / sizeof(INT64));
+        }
+        catch (OutOfMemoryException)
+        {
+            // OK, no cache then
+            return NULL;
+        }
+    }
+
+    TableMask(table) = size - 1;
+
+    // Fibonacci hash reduces the value into desired range by shifting right by the number of leading zeroes in 'size-1' 
+    DWORD bitCnt;
+#if BIT64
+    BitScanReverse64(&bitCnt, size - 1);
+    HashShift(table) = (BYTE)(63 - bitCnt);
+#else
+    BitScanReverse(&bitCnt, size - 1);
+    HashShift(table) = (BYTE)(31 - bitCnt);
+#endif
+
+    return table;
+}
+
+BOOL CastCache::MaybeReplaceCacheWithLarger(DWORD size)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    BASEARRAYREF newTable = CreateCastCache(size);
+    if (!newTable)
+    {
+        return FALSE;
+    }
+
+    // since we are resizing, there must be a cache
+    _ASSERTE(s_cache);
+
+    OBJECTREF currentTableRef = ObjectFromHandle(s_cache);
+    OBJECTREF prevTableRef = (OBJECTREF)(Object*)InterlockedCompareExchangeObjectInHandle(s_cache, newTable, currentTableRef);
+
+    return prevTableRef == currentTableRef;
+}
+
+void CastCache::FlushCurrentCache()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    OBJECTHANDLE cache = s_cache;
+    if (cache == NULL)
+    {
+        cache = CreateGlobalHandle(NULL);
+        OBJECTHANDLE prevCache = FastInterlockCompareExchangePointer(&s_cache, cache, NULL);
+        if (prevCache != NULL)
+        {
+            // someone has already created the cache, use that
+            DestroyGlobalHandle(cache);
+            cache = prevCache;
+        }
+    }
+
+    BASEARRAYREF currentTableRef = (BASEARRAYREF)ObjectFromHandle(cache);
+    int size = !currentTableRef ? INITIAL_CACHE_SIZE : CacheElementCount(currentTableRef);
+
+    BASEARRAYREF newTable = CreateCastCache(size);
+    StoreObjectInHandle(cache, newTable);
+}
+
+TypeHandle::CastResult CastCache::TryGet(TADDR source, TADDR target)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    OBJECTHANDLE cache = s_cache;
+    if (cache == NULL)
```

Comment:
It should be fine to allocate the handle eagerly during startup to avoid this null check.

Reply:
There are also scenarios when we unload/drop the cache. I'll keep the null check here to be safe.

---
In reply to: [307607118](https://github.com/dotnet/coreclr/pull/23548#discussion_r307607118) [](ancestors = 307607118)

Reply:
You should not need to be dropping the handle when you are dropping the cache.

Reply:
I see what you mean. When can I initialize the handle ? I assume initializer of `s_cache` maybe too early to have handles?

---
In reply to: [332115247](https://github.com/dotnet/coreclr/pull/23548#discussion_r332115247) [](ancestors = 332115247)

Reply:
Somewhere in the middle of EEStartupHelper